### PR TITLE
Expand/collapse Topology Application Groups

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Addresses stories:
https://issues.redhat.com/browse/ODC-2459
https://issues.redhat.com/browse/ODC-2548
https://issues.redhat.com/browse/ODC-2460
https://issues.redhat.com/browse/ODC-2623

Fixes:
https://issues.redhat.com/browse/ODC-2420
https://issues.redhat.com/browse/ODC-2780

image

image

image

image

image

image

/cc @serenamarie125 @Veethika @openshift/team-ux-review

### CODE REVIEW

**Comment by christianvogt** [comment]
Screenshot looks good. We'll probably want a follow-up where we can size the groupings better based on the content.

File: frontend/packages/topology/src/behavior/usePanZoom.tsx [hunk 1]
```diff
@@ -31,6 +36,9 @@ export const usePanZoom = (zoomExtent: [number, number] = ZOOM_EXTENT): PanZoomR
         if (node) {
           // TODO fix any type
           const $svg = d3.select(node.ownerSVGElement) as any;
+          if (node && node.ownerSVGElement) {
+            node.ownerSVGElement.addEventListener('mousedown', propagatePanZoomMouseEvent);
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/usePanZoom.tsx]: Make sure to remove this listener in the dispose function.

Comment (jeff-phillips-18) [reply frontend/packages/topology/src/behavior/usePanZoom.tsx]: done

**Comment by jeff-phillips-18** [comment]
Updated to show group badge in application group label:

image

**Comment by andrewballantyne** [comment]
I think your CI tests are bugged (they don't appear to be moving for a few hours now). I know the gpc-console one takes almost 2 hours but the others (images/analyze) do not.

/assign

Hopefully your fix push will rekick it.

File: frontend/packages/dev-console/src/components/topology/Topology.tsx [hunk 2]
```diff
@@ -72,6 +76,12 @@ const Topology: React.FC<TopologyProps> = ({ data, serviceBinding }) => {
     componentFactoryRef.current.serviceBinding = serviceBinding;
   }, [serviceBinding]);
 
+  React.useEffect(() => {
+    action(() => {
+      visRef.current.setTypeCollapsed(TYPE_APPLICATION_GROUP, !filters.display.appGrouping);
+    })();
+  }, [filters.display.appGrouping]);
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/Topology.tsx]: `filters` is apparently not always present? I got a NPE.

**Comment by christianvogt** [comment]
changes to the zoom behavior to fix the click through for hiding dropdown menus looks good.

**Comment by andrewballantyne** [comment]
Ah right, Topology package doesn't have a label and I cannot approve that.

/lgtm

File: frontend/packages/topology/src/elements/BaseNode.ts [hunk 3]
```diff
@@ -130,7 +140,15 @@ export default class BaseNode<E extends NodeModel = NodeModel, D = any> extends
   }
 
   isGroup(): boolean {
-    return this.group;
+    return this.group && !this.collapsed;
+  }
+
+  isCollapsed(): boolean {
+    return this.collapsed || this.getController().getTypeCollapsed(this.getType());
```
Comment (christianvogt) [inline frontend/packages/topology/src/elements/BaseNode.ts]: This property should be set and read as is regardless of some filter. The controller shouldn't have a collapsed type state.thats dev console specific and related to filters.

File: frontend/packages/topology/stories/91-CollapsibleGroups.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+import data from './data/group-types';
+import GroupHull from './components/GroupHull';
+import Group from './components/DefaultGroup';
+import Node from './components/DefaultNode';
+import defaultComponentFactory from './components/defaultComponentFactory';
+import '@patternfly/patternfly/patternfly.css';
+import '@patternfly/patternfly/patternfly-addons.css';
+
+export default {
+  title: 'Collapsible Groups',
```
[the remark this answers is not in the document]
Comment (jeff-phillips-18) [reply frontend/packages/topology/stories/91-CollapsibleGroups.stories.tsx]: Stories are listed in alphabetical order in the storybook. I wanted this one last so...

File: frontend/packages/topology/src/types.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@ import { Padding, Translatable } from './geom/types';
 export type PointTuple = [number, number];
 
 export interface Layout {
-  layout(): void;
+  layout(initialize?: boolean): void;
```
[the remark this answers is not in the document]
Comment (jeff-phillips-18) [reply frontend/packages/topology/src/types.ts]: layout can be called from two places:
-  On the control bar, where we want it to reset locations and layout everything fresh
- After a layout change (collapsing groups) where we want it to keep the current locations where possible.

so, the initialize parameter tells us if the caller wants to initialize the locations or use what we already have.

**Comment by serenamarie125** [comment]
@jeff-phillips-18 this is looking good - couple of questions/comments

1. is this PR supposed to include the Filter which will expand/collapse all apps ?
1. is this PR supposed to include the expand/collapse icon which will expand/collapse individual applications at a time?
1. is this filter setting saved per user per session?  @andrewballantyne @christianvogt unsure if we have a story around this, but if possible we should.  If not, I'll enter a Bug/Enhancement
1. I agree that the visuals of the collapsed grouping should be improved.  We have Sunil working on that, and I think we can make those visual tweaks once this PR is merged.

**Comment by jeff-phillips-18** [comment]
> @jeff-phillips-18 this is looking good - couple of questions/comments
> 
> 1. is this PR supposed to include the Filter which will expand/collapse all apps ?

This includes collapse/expand based on the `Display` dropdown settings

> 2. is this PR supposed to include the expand/collapse icon which will expand/collapse individual applications at a time?

No, this PR does not provide a mechanism to expand/collapse individual items

> 3. is this filter setting saved per user per session?  @andrewballantyne @christianvogt unsure if we have a story around this, but if possible we should.  If not, I'll enter a Bug/Enhancement

The `Display` filters are saved per session

> 4. I agree that the visuals of the collapsed grouping should be improved.  We have Sunil working on that, and I think we can make those visual tweaks once this PR is merged.

**Comment by andrewballantyne** [comment]
@serenamarie125 All display filters are saved on each change. Was handled as a whole and not individually.

File: frontend/packages/topology/src/elements/BaseNode.ts [hunk 6]
```diff
@@ -65,12 +74,6 @@ export default class BaseNode<E extends NodeModel = NodeModel, D = any> extends
     const { padding } = this.getStyle<GroupStyle>();
     const result = rect.padding(padding);
 
-    // ensure size is at least the size of the set bounds
-    result.setSize(
```
[the remark this answers is not in the document]
Comment (jeff-phillips-18) [reply frontend/packages/topology/src/elements/BaseNode.ts]: We are using the bounds for the collapsed size. Making a non-collapsed group this size doesn't make sense anymore. We could add a min-size for groups if we feel this is necessary still.

**Comment by jeff-phillips-18** [comment]
@andrewballantyne @christianvogt  Lots of rework done. PTAL

**Comment by jeff-phillips-18** [comment]
Fixed lint error

**Comment by jeff-phillips-18** [comment]
Removed aggregate edge handling from Visualization, make application defined.

File: frontend/packages/topology/src/elements/BaseNode.ts [hunk 7]
```diff
@@ -133,6 +143,21 @@ export default class BaseNode<E extends NodeModel = NodeModel, D = any> extends
...
+    return this.collapsed;
+  }
+
+  setCollapsed(collapsed: boolean): void {
+    if (collapsed !== this.collapsed) {
+      const prevCenter = this.getBounds()
+        .getCenter()
+        .clone();
+      this.collapsed = collapsed;
+      this.getBounds().setCenter(prevCenter.x, prevCenter.y);
```
Comment (christianvogt) [inline frontend/packages/topology/src/elements/BaseNode.ts]: We should have a TODO with a comment why this is here so we can address it better.

File: frontend/packages/topology/src/elements/BaseNode.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/topology/src/elements/BaseNode.ts]: This was used by knative service and other groups that don't currently have children.
If you kept this, does it mess up anything? You've already added a guard in the bounds getter.

File: frontend/packages/topology/src/behavior/withRemoveConnector.tsx [hunk 8]
```diff
@@ -38,16 +48,36 @@ export const withRemoveConnector = <P extends WithRemoveConnectorProps & Element
...
+
     return (
       <WrappedComponent
         {...(props as any)}
         onShowRemoveConnector={onShowRemoveConnector}
         onHideRemoveConnector={onHideRemoveConnector}
+        removeEdgeRef={removeEdgeRef}
       >
-        {show && renderRemove(props.element, onRemove)}
+        {show && renderRemove(props.element, onRemove, startPoint, endPoint)}
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withRemoveConnector.tsx]: Maybe default to this behavior instead:
```suggestion
        {show && renderRemove(props.element, onRemove, startPoint || props.element.getStartPoint(), endPoint || props.element.getEndPoint())}
```

File: frontend/packages/topology/src/behavior/withRemoveConnector.tsx [hunk 9]
```diff
@@ -10,23 +12,31 @@ type ElementProps = {
...
   onShowRemoveConnector?: () => void;
   onHideRemoveConnector?: () => void;
+  removeEdgeRef?: (svgElement: SVGElement) => void;
 };
 
 type RemoveRenderer = (
   edge: Edge,
   onRemove: (edge: Edge) => void,
+  startPoint?: Point,
+  endPoint?: Point,
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withRemoveConnector.tsx]: Probably useful to make these always present with the defaults being the edge start and end point?

File: frontend/packages/dev-console/src/components/topology/components/edges/ConnectsTo.tsx [hunk 10]
```diff
@@ -31,10 +31,16 @@ const ConnectsTo: React.FC<ConnectsToProps> = ({ element, targetDragRef, childre
...
-  const edgeClasses = classNames('odc-connects-to', { 'odc2-m-editable': editAccess });
+  const edgeClasses = classNames('odc-connects-to', {
+    'odc2-m-editable':
+      editAccess && !element.getSource().isCollapsed() && !element.getTarget().isCollapsed(),
+  });
+
   return (
     <BaseEdge className={edgeClasses} element={element} {...others}>
-      <EdgeConnectorArrow dragRef={editAccess ? targetDragRef : undefined} edge={element} />
+      {(!element.getSource().isCollapsed() || !element.getTarget().isCollapsed()) && (
```
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/topology/components/edges/ConnectsTo.tsx]: Is all this still required now that we have an aggregate edge?

[the remark this answers is not in the document]
Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/components/topology/components/edges/ConnectsTo.tsx]: nope, thanks

File: frontend/packages/topology/src/behavior/withRemoveConnector.tsx (hunk 9 above)
Comment (jeff-phillips-18) [reply frontend/packages/topology/src/behavior/withRemoveConnector.tsx]: when using `withRemoveConnector` we don't yet have the edge.

Comment (jeff-phillips-18) [reply frontend/packages/topology/src/behavior/withRemoveConnector.tsx]: Actually, with the rework, this is no longer needed.

File: frontend/packages/topology/src/elements/BaseNode.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeff-phillips-18) [reply frontend/packages/topology/src/elements/BaseNode.ts]: yeah, it makes the size of the knative service at least as big as the collapsed version which is incorrect.

File: frontend/packages/topology/src/elements/BaseNode.ts (hunk 7 above)
Comment (jeff-phillips-18) [reply frontend/packages/topology/src/elements/BaseNode.ts]: added

**Comment by jeff-phillips-18** [comment]
Fixed the empty knative service display

**Comment by andrewballantyne** [comment]
@jeff-phillips-18 Is the intent to show a collapsed Knative when it's starting up?

image

**Comment by jeff-phillips-18** [comment]
Yes. It's better than an empty grey box but we still have https://issues.redhat.com/browse/ODC-2437 to address it better in the future.

**Comment by andrewballantyne** [comment]
> Yes. It's better than an empty grey box but we still have https://issues.redhat.com/browse/ODC-2437 to address it better in the future.

Hmmm, not sure it is better... just a different kind of misleading 😄

**Comment by spadgett** [comment]
/hold
for merge queue fix #4065

**Pull request merged**